pipeline_kinesis.c

```diff
@@ -0,0 +1,920 @@
+	char *endpoint_region;
+	char *endpoint_credfile;
+	char *endpoint_url;
+
+	char *relation;
+	char *kinesis_stream;
+
+	int batchsize;
+	int num_shards;
+
+	ShardState *shards;
+
+} KinesisConsumerState;
+
+void
+_PG_init(void)
+{
+	HASHCTL ctl;
+
+	MemSet(&ctl, 0, sizeof(HASHCTL));
+
+	ctl.keysize = sizeof(Oid);
+	ctl.entrysize = sizeof(KinesisConsumerInfo);
+	ctl.hash = oid_hash;
+
+	consumer_info = ShmemInitHash("KinsesisConsumerInfo", 4, 64,
+			&ctl, HASH_ELEM | HASH_FUNCTION);
+}
+
+/*
+ * kinesis_add_endpoint
+ *
+ * Add endpoint with name, region, credfile, url
+ */
+PG_FUNCTION_INFO_V1(kinesis_add_endpoint);
+Datum
+kinesis_add_endpoint(PG_FUNCTION_ARGS)
+{
+	char *query = "INSERT INTO pipeline_kinesis_endpoints VALUES ($1, $2, $3, $4)";
+    Oid argtypes[4] = { TEXTOID, TEXTOID, TEXTOID, TEXTOID };
+
+	Datum values[4];
+    char nulls[4];
+
+	MemSet(nulls, 0, sizeof(nulls));
+
+	values[0] = PG_GETARG_DATUM(0);
+	values[1] = PG_GETARG_DATUM(1);
+	values[2] = PG_GETARG_DATUM(2);
+
+	if (PG_ARGISNULL(3))
+		nulls[3] = 'n';
+	else
+		values[3] = PG_GETARG_DATUM(3);
+
+	SPI_connect();
+
+    if (SPI_execute_with_args(query, 4, argtypes,
+				values, nulls, false, 1) != SPI_OK_INSERT)
+		elog(ERROR, "could not add endpoint");
+
+	SPI_finish();
+
+	RETURN_SUCCESS();
+}
+
+
+/*
+ * kinesis_remove_endpoint
+ *
+ * Remove kinesis endpoint matching name
+ */
+PG_FUNCTION_INFO_V1(kinesis_remove_endpoint);
+Datum
+kinesis_remove_endpoint(PG_FUNCTION_ARGS)
+{
+	char *query = "DELETE FROM pipeline_kinesis_endpoints WHERE name = $1;";
+    Oid argtypes[1] = { TEXTOID };
+
+	Datum values[1];
+    char nulls[1];
+
+	MemSet(nulls, 0, sizeof(nulls));
+
+	if (PG_ARGISNULL(0))
+		elog(ERROR, "endpoint cannot be null");
+
+	values[0] = PG_GETARG_DATUM(0);
+
+	SPI_connect();
+
+    if (SPI_execute_with_args(query, 1, argtypes,
+				values, nulls, false, 1) != SPI_OK_DELETE)
+		elog(ERROR, "could not remove endpoint");
+
+	SPI_finish();
+	RETURN_SUCCESS();
+}
+
+volatile int got_sigterm = 0;
+
+static void
+kinesis_consume_main_sigterm(SIGNAL_ARGS)
+{
+	int save_errno = errno;
+
+	got_sigterm = true;
+
+	if (MyProc)
+		SetLatch(&MyProc->procLatch);
+
+	errno = save_errno;
+}
+
+extern void kinesis_consume_main(Datum arg);
+
+static void
+log_fn(void *ctx, const char *s)
+{
+	/* aws log messages have a newline on them, so only log n-1 */
+
+	/* Note - there is a mutex in the cpp caller protecting this */
+	/* it may be more appropriate to put the mutex on this side */
+
+	int n = strlen(s);
+
+	if (n > 0)
+		elog(LOG, "%.*s", n-1, s);
+}
+
+/*
+ * get_copy_statement
+ *
+ * Get the COPY statement that will be used to write messages to a stream
+ *
+ * TODO - support copy formatting options
+ */
+static CopyStmt *
+get_copy_statement(const char *relname)
+{
+	MemoryContext old = MemoryContextSwitchTo(CacheMemoryContext);
+	CopyStmt *stmt = makeNode(CopyStmt);
+	Relation rel;
+	TupleDesc desc;
+	int i;
+
+	RangeVar *rv = makeRangeVar(NULL, (char *) relname, -1);
```

Comment:
The relname could be qualified. Use `makeRangeVarFromNameList(textToQualifiedNameList(relname));`